// @microsoft/rush-lib/lib/pluginFramework/logging/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const node_core_library_1 = require("@rushstack/node-core-library");
class Logger {
    constructor(options) {
        this._errors = [];
        this._warnings = [];
        this._options = options;
        this.terminal = new node_core_library_1.Terminal(options.terminalProvider);
    }
    get errors() {
        return [...this.errors];
    }
    get warnings() {
        return [...this.warnings];
    }
    static getErrorMessage(error) {
        return error.message;
    }
    /**
     * {@inheritdoc ILogger.emitError}
     */
    emitError(error) {
        this._errors.push(error);
        this.terminal.writeErrorLine(`Error: ${Logger.getErrorMessage(error)}`);
        if (this._shouldPrintStacks && error.stack) {
            this.terminal.writeErrorLine(error.stack);
        }
    }
    /**
     * {@inheritdoc ILogger.emitWarning}
     */
    emitWarning(warning) {
        this._warnings.push(warning);
        this.terminal.writeWarningLine(`Warning: ${Logger.getErrorMessage(warning)}`);
        if (this._shouldPrintStacks && warning.stack) {
            this.terminal.writeWarningLine(warning.stack);
        }
    }
    get _shouldPrintStacks() {
        return this._options.getShouldPrintStacks();
    }
}
exports.Logger = Logger;
