// @microsoft/rush-lib/lib/logic/Autoinstaller.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Autoinstaller = void 0;
const safe_1 = __importDefault(require("colors/safe"));
const path = __importStar(require("path"));
const node_core_library_1 = require("@rushstack/node-core-library");
const Utilities_1 = require("../utilities/Utilities");
const node_core_library_2 = require("@rushstack/node-core-library");
const PackageJsonEditor_1 = require("../api/PackageJsonEditor");
const InstallHelpers_1 = require("./installManager/InstallHelpers");
const RushGlobalFolder_1 = require("../api/RushGlobalFolder");
const RushConstants_1 = require("./RushConstants");
const LastInstallFlag_1 = require("../api/LastInstallFlag");
class Autoinstaller {
    constructor(options) {
        this.name = options.autoinstallerName;
        this._rushConfiguration = options.rushConfiguration;
        this._restrictConsoleOutput = options.restrictConsoleOutput || false;
        Autoinstaller.validateName(this.name);
    }
    // Example: .../common/autoinstallers/my-task
    get folderFullPath() {
        return path.join(this._rushConfiguration.commonAutoinstallersFolder, this.name);
    }
    // Example: .../common/autoinstallers/my-task/package-lock.yaml
    get shrinkwrapFilePath() {
        return path.join(this._rushConfiguration.commonAutoinstallersFolder, this.name, this._rushConfiguration.shrinkwrapFilename);
    }
    // Example: .../common/autoinstallers/my-task/package.json
    get packageJsonPath() {
        return path.join(this._rushConfiguration.commonAutoinstallersFolder, this.name, 'package.json');
    }
    static validateName(autoinstallerName) {
        const nameOrError = node_core_library_2.PackageName.tryParse(autoinstallerName);
        if (nameOrError.error) {
            throw new Error(`The specified name "${autoinstallerName}" is invalid: ` + nameOrError.error);
        }
        if (nameOrError.scope) {
            throw new Error(`The specified name "${autoinstallerName}" must not contain an NPM scope`);
        }
    }
    async prepareAsync() {
        const autoinstallerFullPath = this.folderFullPath;
        if (!node_core_library_1.FileSystem.exists(autoinstallerFullPath)) {
            throw new Error(`The autoinstaller ${this.name} does not exist, Please run\nrush init-autoinstaller --name ${this.name}\n`);
        }
        const rushGlobalFolder = new RushGlobalFolder_1.RushGlobalFolder();
        await InstallHelpers_1.InstallHelpers.ensureLocalPackageManager(this._rushConfiguration, rushGlobalFolder, RushConstants_1.RushConstants.defaultMaxInstallAttempts, this._restrictConsoleOutput);
        // Example: common/autoinstallers/my-task/package.json
        const relativePathForLogs = path.relative(this._rushConfiguration.rushJsonFolder, autoinstallerFullPath);
        this._logIfConsoleOutputIsNotRestricted(`Acquiring lock for "${relativePathForLogs}" folder...`);
        const lock = await node_core_library_1.LockFile.acquire(autoinstallerFullPath, 'autoinstaller');
        // Example: .../common/autoinstallers/my-task/.rush/temp
        const lastInstallFlagPath = path.join(autoinstallerFullPath, RushConstants_1.RushConstants.projectRushFolderName, 'temp');
        const packageJsonPath = path.join(autoinstallerFullPath, 'package.json');
        const packageJson = node_core_library_1.JsonFile.load(packageJsonPath);
        const lastInstallFlag = new LastInstallFlag_1.LastInstallFlag(lastInstallFlagPath, {
            node: process.versions.node,
            packageManager: this._rushConfiguration.packageManager,
            packageManagerVersion: this._rushConfiguration.packageManagerToolVersion,
            packageJson: packageJson,
            rushJsonFolder: this._rushConfiguration.rushJsonFolder
        });
        if (!lastInstallFlag.isValid() || lock.dirtyWhenAcquired) {
            // Example: ../common/autoinstallers/my-task/node_modules
            const nodeModulesFolder = path.join(autoinstallerFullPath, 'node_modules');
            if (node_core_library_1.FileSystem.exists(nodeModulesFolder)) {
                this._logIfConsoleOutputIsNotRestricted('Deleting old files from ' + nodeModulesFolder);
                node_core_library_1.FileSystem.ensureEmptyFolder(nodeModulesFolder);
            }
            // Copy: .../common/autoinstallers/my-task/.npmrc
            Utilities_1.Utilities.syncNpmrc(this._rushConfiguration.commonRushConfigFolder, autoinstallerFullPath);
            this._logIfConsoleOutputIsNotRestricted(`Installing dependencies under ${autoinstallerFullPath}...\n`);
            Utilities_1.Utilities.executeCommand({
                command: this._rushConfiguration.packageManagerToolFilename,
                args: ['install', '--frozen-lockfile'],
                workingDirectory: autoinstallerFullPath,
                keepEnvironment: true
            });
            // Create file: ../common/autoinstallers/my-task/.rush/temp/last-install.flag
            lastInstallFlag.create();
            this._logIfConsoleOutputIsNotRestricted('Auto install completed successfully\n');
        }
        else {
            this._logIfConsoleOutputIsNotRestricted('Autoinstaller folder is already up to date\n');
        }
        lock.release();
    }
    update() {
        const autoinstallerPackageJsonPath = path.join(this.folderFullPath, 'package.json');
        if (!node_core_library_1.FileSystem.exists(autoinstallerPackageJsonPath)) {
            throw new Error(`The specified autoinstaller path does not exist: ` + autoinstallerPackageJsonPath);
        }
        this._logIfConsoleOutputIsNotRestricted(`Updating autoinstaller package: ${autoinstallerPackageJsonPath}`);
        let oldFileContents = '';
        if (node_core_library_1.FileSystem.exists(this.shrinkwrapFilePath)) {
            oldFileContents = node_core_library_1.FileSystem.readFile(this.shrinkwrapFilePath, { convertLineEndings: node_core_library_1.NewlineKind.Lf });
            this._logIfConsoleOutputIsNotRestricted('Deleting ' + this.shrinkwrapFilePath);
            node_core_library_1.FileSystem.deleteFile(this.shrinkwrapFilePath);
        }
        // Detect a common mistake where PNPM prints "Already up-to-date" without creating a shrinkwrap file
        const packageJsonEditor = PackageJsonEditor_1.PackageJsonEditor.load(this.packageJsonPath);
        if (packageJsonEditor.dependencyList.length === 0 && packageJsonEditor.dependencyList.length === 0) {
            throw new Error('You must add at least one dependency to the autoinstaller package' +
                ' before invoking this command:\n' +
                this.packageJsonPath);
        }
        this._logIfConsoleOutputIsNotRestricted();
        Utilities_1.Utilities.syncNpmrc(this._rushConfiguration.commonRushConfigFolder, this.folderFullPath);
        Utilities_1.Utilities.executeCommand({
            command: this._rushConfiguration.packageManagerToolFilename,
            args: ['install'],
            workingDirectory: this.folderFullPath,
            keepEnvironment: true
        });
        this._logIfConsoleOutputIsNotRestricted();
        if (this._rushConfiguration.packageManager === 'npm') {
            this._logIfConsoleOutputIsNotRestricted(safe_1.default.bold('Running "npm shrinkwrap"...'));
            Utilities_1.Utilities.executeCommand({
                command: this._rushConfiguration.packageManagerToolFilename,
                args: ['shrinkwrap'],
                workingDirectory: this.folderFullPath,
                keepEnvironment: true
            });
            this._logIfConsoleOutputIsNotRestricted('"npm shrinkwrap" completed');
            this._logIfConsoleOutputIsNotRestricted();
        }
        if (!node_core_library_1.FileSystem.exists(this.shrinkwrapFilePath)) {
            throw new Error('The package manager did not create the expected shrinkwrap file: ' + this.shrinkwrapFilePath);
        }
        const newFileContents = node_core_library_1.FileSystem.readFile(this.shrinkwrapFilePath, {
            convertLineEndings: node_core_library_1.NewlineKind.Lf
        });
        if (oldFileContents !== newFileContents) {
            this._logIfConsoleOutputIsNotRestricted(safe_1.default.green('The shrinkwrap file has been updated.') + '  Please commit the updated file:');
            this._logIfConsoleOutputIsNotRestricted(`\n  ${this.shrinkwrapFilePath}`);
        }
        else {
            this._logIfConsoleOutputIsNotRestricted(safe_1.default.green('Already up to date.'));
        }
    }
    _logIfConsoleOutputIsNotRestricted(message) {
        if (!this._restrictConsoleOutput) {
            console.log(message);
        }
    }
}
exports.Autoinstaller = Autoinstaller;
